File: src/main/java/com/example/rpc/client/PingClient.java
package com.example.rpc.client;

import com.example.rpc.services.PingRequest;
import com.example.rpc.services.PingResponse;
import com.example.rpc.services.PingServiceGrpc;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import org.springframework.stereotype.Service;

@Service
public class PingClient {

    public static void main(String[] args) {
        System.out.println("Manual client received response: " + new PingClient().sendPing());
    }

    public String sendPing() {
        ManagedChannel channel = ManagedChannelBuilder.forAddress("127.0.0.1", 50051)
                .usePlaintext()
                .build();

        PingServiceGrpc.PingServiceBlockingStub stub = PingServiceGrpc.newBlockingStub(channel);
        PingRequest request = PingRequest.newBuilder()
                .setMessage("Ping! [springboot client]")
                .build();

        PingResponse response = stub.getPing(request);
        System.out.println("Java client received response: " + response.getMessage());
        channel.shutdown();
        return response.getMessage();
    }
}

File: src/main/python/ping_client.py
import grpc
import ping_pb2
import ping_pb2_grpc

def run():
    channel = grpc.insecure_channel('localhost:50051')
    stub = ping_pb2_grpc.PingServiceStub(channel)

    request = ping_pb2.PingRequest(message='Ping! [python client]')

    response = stub.getPing(request)
    print("Received response:", response.message)

if __name__ == '__main__':
    run()